### PR TITLE
mds/tools: add list volumes on majority-peers-offline copysets

### PR DESCRIPTION
**Pull request by bai-charisu**
Change-Id: I0bf2dbe171cab2080b05de6ea2bc0c59bd6c71ff

### What problem does this PR solve?

curve_ops_tool support list all volumes on majority-peers-offline copysets.

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by bai-charisu** [comment]
recheck

**Comment by bai-charisu** [comment]
recheck

File: src/tools/mds_client.cpp [hunk 1]
```diff
@@ -327,6 +327,40 @@ int MDSClient::CreateFile(const std::string& fileName, uint64_t length) {
...
+        *copysetPtr = copyset;
+    }
+    curve::mds::CurveFSService_Stub stub(&channel_);
+
+    void (curve::mds::CurveFSService_Stub::*fp)(
+                            google::protobuf::RpcController*,
+                            const curve::mds::ListVolumesOnCopysetsRequest*,
+                            curve::mds::ListVolumesOnCopysetsResponse*,
+                            google::protobuf::Closure*);
+    fp = &curve::mds::CurveFSService_Stub::ListVolumesOnCopysets;
```
Comment (wu-hanqing) [inline src/tools/mds_client.cpp:346]: `auto fp = &curve::mds::CurveFSService_Stub::ListVolumesOnCopysets;` ?

File: src/mds/nameserver2/curvefs.cpp [hunk 2]
```diff
@@ -1971,6 +1974,77 @@ StatusCode CurveFS::FindFileMountPoint(
...
+                        std::vector<std::string>* fileNames) {
+    std::vector<FileInfo> files;
+    StatusCode ret = ListAllFiles(ROOTINODEID, &files);
+    if (ret != StatusCode::kOK) {
+        LOG(ERROR) << "List all files in root directory fail";
+        return ret;
+    }
+    std::map<LogicalPoolIdType, std::set<CopySetIdType>> copysetMap;
+    for (const auto& copyset : copysets) {
+        if (copysetMap.find(copyset.logicalpoolid()) == copysetMap.end()) {
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/nameserver2/curvefs.cpp]: done

File: src/mds/nameserver2/curvefs.cpp [hunk 3]
```diff
@@ -1971,6 +1974,77 @@ StatusCode CurveFS::FindFileMountPoint(
...
+            return StatusCode::kStorageError;
+        }
+        bool found = false;
+        for (const auto& segment : segments) {
+            if (copysetMap.find(segment.logicalpoolid()) == copysetMap.end()) {
+                continue;
+            }
+            for (int i = 0; i < segment.chunks_size(); i++) {
+                auto copysetId = segment.chunks(i).copysetid();
+                if (copysetMap[segment.logicalpoolid()].find(copysetId)
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/nameserver2/curvefs.cpp]: done

File: src/mds/nameserver2/namespace_service.cpp [hunk 4]
```diff
@@ -1787,6 +1787,39 @@ void NameSpaceService::FindFileMountPoint(
...
+    if (retCode != StatusCode::kOK)  {
+        response->set_statuscode(retCode);
+        LOG(ERROR) << "logid = " << cntl->log_id()
+            << ", ListAllVolumesOnCopysets fail, statusCode = "
+            << retCode << ", StatusCode_Name = " << StatusCode_Name(retCode);
+        return;
+    }
+    response->set_statuscode(StatusCode::kOK);
+    for (const auto& fileName : fileNames) {
+        std::string* name = response->add_filenames();
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/nameserver2/namespace_service.cpp]: done，改为使用set_filenames(string)了

File: src/tools/mds_client.cpp (hunk 1 above)
Comment (bai-charisu) [reply src/tools/mds_client.cpp]: done

**Comment by bai-charisu** [comment]
recheck

**Comment by bai-charisu** [comment]
recheck

File: curve-chunkserver/home/nbs/chunkserver_deploy.sh [hunk 5]
```diff
@@ -296,7 +296,6 @@ function deploy_one {
   curve-format -allocatePercent=80 \
   -filePoolDir=$dirname/chunkfilepool \
   -filePoolMetaPath=$dirname/chunkfilepool.meta \
-  -fileSize=8388608 \
```
Comment (wu-hanqing) [inline curve-chunkserver/home/nbs/chunkserver_deploy.sh:299]: 这个为什么去掉了？

Comment (bai-charisu) [reply curve-chunkserver/home/nbs/chunkserver_deploy.sh:299]: 这里应该是16M的，默认就是16M

File: src/mds/nameserver2/namespace_service.cpp [hunk 6]
```diff
@@ -1787,6 +1787,38 @@ void NameSpaceService::FindFileMountPoint(
...
+        LOG(ERROR) << "logid = " << cntl->log_id()
+            << ", ListAllVolumesOnCopysets fail, statusCode = "
+            << retCode << ", StatusCode_Name = " << StatusCode_Name(retCode);
+        return;
+    }
+    response->set_statuscode(StatusCode::kOK);
+    for (const auto& fileName : fileNames) {
+        response->add_filenames(fileName);
+    }
+    LOG(INFO) << "logid = " << cntl->log_id()
```
Comment (ilixiaocui) [inline src/mds/nameserver2/namespace_service.cpp:1818]: 打印下文件数量

File: src/tools/copyset_check.cpp [hunk 7]
```diff
@@ -222,6 +225,8 @@ void CopysetCheck::PrintHelp(const std::string& command) {
         std::cout << "curve_ops_tool check-operator -opName=" << kTotalOpName
                   << "/" << kChangeOpName << "/" << kAddOpName << "/"
                   << kRemoveOpName << "/" << kTransferOpName << std::endl;
+    } else if (command == kListMajorityOfflineVolumes) {
```
Comment (ilixiaocui) [inline src/tools/copyset_check.cpp:228]: ListMayBrokenVolumes?

File: src/tools/copyset_check.cpp [hunk 8]
```diff
@@ -322,5 +327,19 @@ void CopysetCheck::PrintExcepChunkservers() {
...
 }
+
+int CopysetCheck::PrintMajorityOfflineVolumes() {
+    std::vector<std::string> fileNames;
+    int res = core_->ListMajorityOfflineVolumes(&fileNames);
+    if (res != 0) {
+        std::cout << "ListMajorityOfflineVolumes fail" << std::endl;
+        return -1;
+    }
+    for (const auto& fileName : fileNames) {
```
Comment (ilixiaocui) [inline src/tools/copyset_check.cpp:338]: total count:

File: src/tools/copyset_check_core.cpp [hunk 9]
```diff
@@ -828,5 +828,57 @@ void CopysetCheckCore::Clear() {
...
+    std::vector<common::CopysetInfo> copysets;
+    for (auto iter = copysets_[kMajorityPeerNotOnline].begin();
+                    iter != copysets_[kMajorityPeerNotOnline].end(); ++iter) {
+        std::string gid = *iter;
+        uint64_t groupId;
+        if (!curve::common::StringToUll(gid, &groupId)) {
+            std::cout << "parse group id fail: " << groupId << std::endl;
+            continue;
+        }
+        PoolIdType lgId = groupId >> 32;
```
Comment (ilixiaocui) [inline src/tools/copyset_check_core.cpp:852]: 应该有groupid to copysetid的函数

File: src/tools/copyset_check_core.h [hunk 10]
```diff
@@ -229,6 +229,15 @@ class CopysetCheckCore {
     */
     virtual bool CheckChunkServerOnline(const std::string& chunkserverAddr);
 
+    /**
+    * @brief 列出大多数副本挂掉的copyset上面的卷
```
Comment (ilixiaocui) [inline src/tools/copyset_check_core.h:233]: in english

File: src/tools/copyset_check_core.h [hunk 11]
```diff
@@ -282,6 +291,11 @@ class CopysetCheckCore {
                                    const std::set<std::string>& groupIds,
                                    bool queryLeader = true);
 
+    /**
+    * @brief 检查offline的chunkserver上面的copyset
```
Comment (ilixiaocui) [inline src/tools/copyset_check_core.h:295]: in english

File: src/mds/nameserver2/curvefs.cpp [hunk 12]
```diff
@@ -1971,6 +1974,72 @@ StatusCode CurveFS::FindFileMountPoint(
     return StatusCode::kFileNotExists;
 }
 
+StatusCode CurveFS::ListVolumesOnCopyset(
+                        const std::vector<common::CopysetInfo>& copysets,
+                        std::vector<std::string>* fileNames) {
+    std::vector<FileInfo> files;
+    StatusCode ret = ListAllFiles(ROOTINODEID, &files);
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/nameserver2/curvefs.cpp:1981]: etcdclient暂时没有实现分批获取file

File: curve-ansible/roles/install_package/templates/chunkserver_deploy.sh.j2 [hunk 13]
```diff
@@ -271,10 +271,11 @@ function deploy_one {
   echo "uuid=$uuid" > $dirname/disk.meta
   echo "uuidmd5=$uuidmd5" >> $dirname/disk.meta
   #格式化chunkfile pool
+
   curve-format -allocatePercent={{ chunk_alloc_percent }} \
   -filePoolDir=$dirname/chunkfilepool \
   -filePoolMetaPath=$dirname/chunkfilepool.meta \
-  -fileSize={{ chunkserver_walfilepool_segment_size }} \
+  -fileSize={{ chunk_size }} \
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply curve-ansible/roles/install_package/templates/chunkserver_deploy.sh.j2:278]: done

File: src/mds/nameserver2/namespace_service.cpp (hunk 6 above)
Comment (bai-charisu) [reply src/mds/nameserver2/namespace_service.cpp:1818]: done

File: src/tools/copyset_check.cpp (hunk 7 above)
Comment (bai-charisu) [reply src/tools/copyset_check.cpp]: done

File: src/tools/copyset_check.cpp (hunk 8 above)
Comment (bai-charisu) [reply src/tools/copyset_check.cpp:338]: done

File: src/tools/copyset_check_core.cpp (hunk 9 above)
Comment (bai-charisu) [reply src/tools/copyset_check_core.cpp]: done

File: src/tools/copyset_check_core.h (hunk 10 above)
Comment (bai-charisu) [reply src/tools/copyset_check_core.h]: done

File: src/tools/copyset_check_core.h (hunk 11 above)
Comment (bai-charisu) [reply src/tools/copyset_check_core.h]: done

File: src/tools/copyset_check.cpp [hunk 14]
```diff
@@ -98,6 +99,8 @@ int CopysetCheck::RunCommand(const std::string& command) {
             return -1;
         }
         return CheckOperator(FLAGS_opName);
+    } else if (command == kListMayBrokenVolumes) {
+        return PrintMajorityOfflineVolumes();
```
Comment (ilixiaocui) [inline src/tools/copyset_check.cpp:103]: 函数名称也相应改一下？

File: src/tools/curve_tool_main.cpp [hunk 15]
```diff
@@ -56,6 +56,7 @@ const char* kHelpStr = "Usage: curve_ops_tool [Command] [OPTIONS...]\n"
         "check-copyset : check the health state of one copyset\n"
         "check-server : check the health state of the server\n"
         "check-operator : check the operators\n"
+        "list-majoroff-vol: list all volumes on majority offline copysets\n"
```
Comment (ilixiaocui) [inline src/tools/curve_tool_main.cpp:59]: 遗漏了

File: src/tools/mds_client.h [hunk 16]
```diff
@@ -174,6 +174,16 @@ class MDSClient {
      */
     virtual int CreateFile(const std::string& fileName, uint64_t length);
 
+    /**
+     *  @brief 列出copyset上面的所有卷
+     *  @param copysets copyset列表
+     *  @param[out] fileNames 文件列表
```
Comment (ilixiaocui) [inline src/tools/mds_client.h:180]: 英文？

File: src/tools/copyset_check.cpp (hunk 14 above)
Comment (bai-charisu) [reply src/tools/copyset_check.cpp]: done

File: src/tools/curve_tool_main.cpp (hunk 15 above)
Comment (bai-charisu) [reply src/tools/curve_tool_main.cpp]: done

File: src/tools/mds_client.h (hunk 16 above)
Comment (bai-charisu) [reply src/tools/mds_client.h]: done

**Review by wu-hanqing — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**